### PR TITLE
Support request and response hooks for Django instrumentation

### LINKED ISSUE #138: Add hook to set custom span name in django instrumentation
Similar to [wsgi](https://github.com/open-telemetry/opentelemetry-python/blob/master/instrumentation/opentelemetry-instrumentation-wsgi/src/opentelemetry/instrumentation/wsgi/__init__.py#L210), accept a callback function in `DjangoInstrumentor().instrument` to be able to pass custom span name.

The [specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#name) say "MAY" for this feature, but we should be consistent across our instrumentations.

Part of https://github.com/open-telemetry/opentelemetry-python/issues/1297

### PR DESCRIPTION
**Pull request by owais**
# Description

Add request and response hooks support for Django instrumentation. This allows users to customize a bunch of things. Users can set custom span names, add additional attributes or other information to spans, inject tracing information into responses, etc.

Fixes #138 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Added unit test

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/__init__.py [hunk 1]
```diff
@@ -11,6 +11,67 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+"""
```
Comment (owais) [inline instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/__init__.py:14]: Opportunistic update. Django was missing docs so copied from README.rst

**Comment by lonewolf3739** [comment]
We have couple of instrumentations where we let users set custom span name with `name_callback`. This change covers it along with many other things one can do. So we will be removing the span name callback?

**Comment by owais** [comment]
I'd like to remove them at don't point yes but not sure but much breakage it'll cause. Let's discuss that in the SIG meeting.

**Comment by lzchen** [comment]
Fixes https://github.com/open-telemetry/opentelemetry-python-contrib/issues/138

**Review by lonewolf3739 — APPROVED** [review]
LGTM

**Review by lzchen — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Pull request merged**